let Product,User,Color;_404‍.x([["createProduct",()=>createProduct],["updateProduct",()=>updateProduct],["deleteProduct",()=>deleteProduct],["getProducts",()=>getProducts]]);_404‍.w("../models",[["Product",["Product"],function(v){Product=v}],["User",["User"],function(v){User=v}],["Color",["Color"],function(v){Color=v}]]);

       const createProduct = async (req, res) => {
  try {
    const product = await Product.create({
      ...req.body,
      user_id: req.user.id,
    });
    const productTobeAssignColors = await Product.findOne({
      where: {
        id: product.id
      },
      include: {
        model: Color,
        as: 'products'
      }
    })
    await productTobeAssignColors.addProducts(req.body.colors, {through: 'ProductColors'})
    // await productTobeAssignColors.setProducts(req.body.colors, { through: 'ProductColors' } ) // many to many update
    // await productTobeAssignColors.removeProducts(req.body.colors, { through: 'ProductColors' } ) // many to many delete

    return res.status(200).json({message: "product is created!", data: product})
  } catch (err) {
    _404‍.g.console.log("error", err)
    res.status(500).json({
      message: 'Something went wrong!'
    })
  }
};

       const updateProduct = async (req, res) => {
  try {
   let user =  await User.findOne()
    let product = await Product.findOne()

    if(user){
      _404‍.g.console.log("dgh;dfgoagfdaf",user)
      await Product.update(req.body, {
        where: {id: product.id, user_id:user.user_id}
      });
      return res.status(200).json({message: "product is updated!"})
    }else{
      console.log("error  somethiong else")
    }

  } catch (err) {
    _404‍.g.console.log("error", err)
    res.status(500).json({
      message: 'Something went wrong!'
    })
  }
};

       const deleteProduct = async (req, res) => {
  try {
    let x = await Product.destroy({
      where: {id: req.params.id}
    })
    return res.status(200).json({message: "product is deleted!"})
  } catch (err) {
    _404‍.g.console.log("error", err)
  }
}

       const getProducts = async (req, res) => {
  try {
    let products = await Product.findAll();
    let users = await User.findAll({
      attributes: {
        exclude: ["password"]
      }
    })
    return res.status(200).json({products: products, users})
  } catch (err) {
    _404‍.g.console.log("error", err)
  }
}